### ISSUE TITLE
Attributes declared as iterables are replaced in the instances by `pydantic-core` `ValidatorIterator` instance

### ISSUE THREAD

**Issue by andresliszt** [pydantic/pydantic#9541]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

When a field is declared as Iterable, the field in the instances is replaced by `ValidatorIterator`. I understand that validators in generators are intended to act once the data is accessed, and that is what `ValidatorIterator` does, but I think it is unintuitive to replace the field in the instance as an instance of this validator. This is more general than what is stated in this issue and is not limited only to `TypedDict` https://github.com/pydantic/pydantic/issues/9467

```python
from typing import Iterable

from pydantic import BaseModel


class MyModel(BaseModel):
    x: Iterable[int]

>>> model = MyModel(x = [1,2,3])
>>> model.x
# ValidatorIterator(index=0, schema=Some(Int(IntValidator { strict: false })))
>>> type(model.x)
<class 'pydantic_core._pydantic_core.ValidatorIterator'>
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.8.0a1
        pydantic-core version: 2.18.3
          pydantic-core build: profile=release pgo=false
                 install path: /somepath/projects/pydantic/pydantic
               python version: 3.10.10 (v3.10.10:aad5f6a891, Feb  7 2023, 08:47:40) [Clang 13.0.0 (clang-1300.0.29.30)]
                     platform: macOS-12.4-x86_64-i386-64bit
             related packages: typing_extensions-4.10.0
                       commit: e67524c0
```

**Comment by afdw** [comment]
See also #9266.

**Comment by sydney-runkle** [comment]
Indeed, this is not ideal. Thanks for the report. I've added #9467 to our v2.8.0 milestone, so we can try to address this as well at that point.

**Comment by mikeleppane** [comment]
Hey,

I dip my toes into this. This line of code looks suspicious:
https://github.com/pydantic/pydantic-core/blob/9507a28f93da317dbfd90a32ec9e66b1c3c0b964/src/validators/generator.rs#L89

In general, I'm surprised to see that these validating methods return some domain object. From a narrow analysis, you could identify two different domains here. One, which validates input and only does that, and one, which builds the correct object. Now, they twine together.

**Comment by sydney-runkle** [comment]
I'm not aware of a workaround at the moment, other than using a custom validator...

**Comment by korroktheslavemaster** [comment]
I just need to serialize the object, so i tried `model_dump_json` which works but somehow mutates the original object. To expand on the previous example:

```python
from typing import Iterable

from pydantic import BaseModel


class MyModel(BaseModel):
  x: Iterable[int]
  y: int
  
model = MyModel(x = [1,2,3], y=4)

>>> model.x
ValidatorIterator(index=0, schema=Some(Int(IntValidator { strict: false })))
>>> model.model_dump_json()
'{"x":[1,2,3],"y":4}'
>>> model.model_dump_json()
'{"x":[],"y":4}'
```
I'm sure this can't be the intended behavior?

**Comment by andresliszt** [comment]
I think is because it's treated as a generator, so once you consume it (first model dump) is not available anymore

**Comment by andresliszt** [comment]
I'd recommend perhaps using a less generic type like `Sequence`

```python
from typing import Sequence

from pydantic import BaseModel


class MyModel(BaseModel):
  x: Sequence[int]
  y: int

model = MyModel(x = [1,2,3], y=4)
print(model.x)
print(model.model_dump_json())
print(model.model_dump_json())
[1, 2, 3]
{"x":[1,2,3],"y":4}
{"x":[1,2,3],"y":4}



```

**Comment by Ravencentric** [comment]
Here's my current workaround, I'm well aware my type isn't as wide as the `collections.abc.Iterable` but in my case I only really needed to cover the commonly used iterables
```py
>>> from typing import Annotated, TypeAlias, TypeVar, Union
>>> from pydantic import BaseModel
>>>
>>> T = TypeVar("T")
>>> Iterable: TypeAlias = Union[set[T], tuple[T, ...], list[T]]
>>>
>>> class Test(BaseModel):
...     a: Iterable[str]
...     b: Iterable[int]
...
>>> test = Test(a=["1", "2"], b=("1", "2"))
>>> test
Test(a=['1', '2'], b={1, 2})
>>> test.model_dump()
{'a': ['1', '2'], 'b': {1, 2}}
>>> test.model_dump_json()
'{"a":["1","2"],"b":[1,2]}'
```

**Comment by sydney-runkle** [comment]
xref https://github.com/pydantic/pydantic/issues/9467#issuecomment-2442097291

**Comment by sarahboufelja** [comment]
Any update on this issue pls?

**Comment by black-snow** [comment]
Just stumbled over this and it breaks my neck.

**Comment by davidhewitt** [comment]
I was looking at this today, and I think that using `ValidatorIterator` to wrap `Iterable` is plain wrong in multiple ways:
- `Iterable` and `Iterator` are different concepts, the former can potentially be iterated multiple times and but the latter can really only be iterated once
- validating `Iterable` should not need to be lazy, because an `Iterable` can potentially be consumed multiple times

I think there are the following ways that pydantic can improve:
- pydantic should support `Iterator` and maybe even `Generator` type annotations, and those types can be wrapped by the lazy `ValidatorIterator` type
- pydantic should NOT wrap `Iterable` types in `ValidatorIterator` and validate them lazily, instead it should iterate and validate immediately, probably storing the result of iteration & validation as a `list` or `tuple`
  - this would be breaking, so sadly for V2 we probably need to deprecate the lazy validation of `Iterable` (encouraging users to migrate to `Iterator` if they actually want lazy validation) and have a config flag to accept the new behaviour now

**Comment by RobertCraigie** [comment]
Any updates here?

**Comment by Viicos** [comment]
@RobertCraigie see https://github.com/pydantic/pydantic-core/pull/1792#issuecomment-3373277959.
